src/cockpit/router.py

```diff
@@ -26,12 +27,47 @@
 logger = logging.getLogger(__name__)
 
 
+class ExecutionQueue(collections.deque):
+    def run(self):
+        for method, args in self:
+            method(*args)
+
+    def wrap(self, method):
+        """Temporarily delay calls to a given class method.
+
+        Functions by replacing the named function at the instance __dict__
+        level, effectively providing an override for exactly one instance.
+        Queues the invocations.  Run them later with .run(). Reverse this call
+        by deleting the named methods from the instance.
+        """
+        # NB: this function is stored in the instance dict and therefore
+        # doesn't function as a descriptor, isn't a method, doesn't get bound,
+        # and therefore doesn't receive a self parameter
+        setattr(method.__self__, method.__func__.__name__, lambda *args: self.append((method, args)))
+
+
 class Endpoint:
     router: Router
+    queue: Optional[ExecutionQueue] = None
 
     def __init__(self, router: Router):
         self.router = router
 
+    def freeze_endpoint(self):
+        self.queue = ExecutionQueue()
+        self.queue.wrap(self.do_channel_control)
+        self.queue.wrap(self.do_channel_data)
+
+    def thaw_endpoint(self):
+        queue = self.queue
+
+        if queue is not None:
```

Comment:
This should become an assertion, at least until we find a situtation where we really can't avoid an unbalanced freeze/thaw.

Reply:
I think I agree, ya.

Reply:
Curiously that fails: Adding `assert self.queue` at the top triggers the assertion, which means that the unit tests somehow run into the `init` event before the `connected` event? I.e. they try to thaw before a freeze?